csharp/src/Apache.Arrow/Arrays/BinaryArray.cs

```diff
@@ -193,10 +194,12 @@ public int GetValueLength(int index)
 
         public ReadOnlySpan<byte> GetBytes(int index)
         {
-            var offset = GetValueOffset(index);
-            var length = GetValueLength(index);
+            if (index < 0 || index >= Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
```

Comment:
The check is duplicated of the check in `GetValueLength`.

How about this?

```csharp
var offset = ValueOffsets[index];
var length = GetValueLength(index);
return ValueBuffer.Span.Slice(offset, length);
```

Reply:
Initially, I implemented as below to avoid duplication of checks.
However, I thought the intention was a little difficult to understand, so it was implemented like current.

```cs
  var length = GetValueLength(index);
  return ValueBuffer.Span.Slice(ValueOffsets[index], length);
```

Also, if we don't care about the type of exception, we can simply remove the check.
In that case, this method throws `ArgumentOutOfRangeException` which ValueOffsets[index] throws.

Reply:
I understand.
How about adding a new helper private method to validate index:

```csharp
private void ValidateIndex(int index)
{
  if (index < 0 || index >= Length)
  {
    throw new ArgumentOutOfRangeException(nameof(index));
  }
}
```

and use it in `GetValueLength` and `GetBytes`?

```csharp
ValidateIndex(index);
var offsets = ValueOffsets;
var offset = offsets[index];
var length = offsets[index + 1] - offset;
return ValueBuffer.Span.Slice(offset, length);
```

@eerhardt What do you think about this case?

Reply:
> Also, if we don't care about the type of exception, we can simply remove the check.
In that case, this method throws IndexOutOfRangeException which ValueOffsets[index] throws.

Typically we do care about the type of exception. Bubbling up an `IndexOutOfRangeException` looks like a bug in our library - similar to if you let something `NullReferenceException`. See the [Design Guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/using-standard-exception-types#nullreferenceexception-indexoutofrangeexception-and-accessviolationexception) for more info about this.

Instead, it is better to throw an `ArgumentOutOfRangeException`.

I think a helper like `ValidateIndex` makes the most sense. Also note that as currently written you are validating the index twice - once in `GetBytes` and then again when `GetBytes` calls `GetValueLength`. Not a huge issue, just something I noticed.